### ISSUE TITLE
Should JSON output be structured with consistent keys?

### ISSUE THREAD

**Issue by alexanderperlis** [lm-sensors/lm-sensors#494]
The current JSON output from `sensors -j` could be improved. The current JSON structure is an object whose keys are *arbitrary* chip labels, whose values in turn are objects whose keys are *arbitrary* feature labels, whose values in turn are objects whose keys are subfeature labels containing *arbitrary* sequence numbers (which come from file names inside `/sys/class/hwmon/*` where the sequence numbers seem to exist merely to avoid name clashes within a single directory — they serve no purpose in the JSON structure). Example of current JSON output:

```json
{
  "BAT1-acpi-0":{
    "Adapter": "ACPI interface",
    "in0":{"in0_input": 16.583},
    "curr1":{"curr1_input": 0.001}
  },
  "bnxt_en-pci-3100":{
    "Adapter": "PCI adapter",
    "temp1":{"temp1_input": 45.000}
  },
  "coretemp-isa-0000":{
    "Adapter": "ISA adapter",
    "Package id 0":{"temp1_input": 24.000, "temp1_max": 93.000},
    "Core 0":{"temp2_input": 21.000, "temp2_max": 93.000},
    "Core 12":{"temp3_input": 20.000, "temp3_max": 93.000}
  },
  "bnxt_en-pci-3101":{
     "Adapter": "PCI adapter",
     "temp1":{"temp1_input": 45.000}
  },
  "coretemp-isa-0001":{
     "Adapter": "ISA adapter",
     "Package id 1":{"temp1_input": 24.000, "temp1_max": 93.000},
     "Core 0":{"temp2_input": 18.000, "temp2_max": 93.000},
     "Core 12":{"temp3_input": 20.000, "temp3_max": 93.000}
  }
}
```

This structure seems hard to work with. For example, how could a consumer easily compute say the “average core temperature”? One would have to look for objects with keys matching `^Core ` and then gather subkeys matching `temp[0-9]+_input`. But languages and tools that work with JSON tend to be designed around fixed key names known ahead of time — their matching capabilities are usually applied to *values* rather than to *keys*. (One can overcome this difficulty by jumping through additional hoops: languages often provide ways to transform keys into values, process those values, and then optionally transform back to keys.)

At a minimum, I contend the JSON structure should be an *array* of chip objects with *consistent* keys, containing an *array* of feature objects with *consistent* keys, in particular with arbitrary sequence numbers removed from the subfeatures. The prior example would become:

```json
[
  {
    "chip_label": "BAT1-acpi-0",
    "adapter": "ACPI interface",
    "features": [
      { "label": "in0", "in_input": 16.583 },
      { "label": "curr1", "curr_input": 0.001 }
    ]
  },
  {
    "chip_label": "bnxt_en-pci-3100",
    "adapter": "PCI adapter",
    "features": [
      { "label": "temp1", "temp_input": 45 }
    ]
  },
  {
    "chip_label": "coretemp-isa-0000",
    "adapter": "ISA adapter",
    "features": [
      { "label": "Package id 0", "temp_input": 24, "temp_max": 93 },
      { "label": "Core 0", "temp_input": 21, "temp_max": 93 },
      { "label": "Core 12", "temp_input": 20, "temp_max": 93 }
    ]
  },
  {
    "chip_label": "bnxt_en-pci-3101",
    "adapter": "PCI adapter",
    "features": [
      { "label": "temp1", "temp_input": 45 }
    ]
  },
  {
    "chip_label": "coretemp-isa-0001",
    "adapter": "ISA adapter",
    "features": [
      { "label": "Package id 1", "temp_input": 24, "temp_max": 93 },
      { "label": "Core 0", "temp_input": 18, "temp_max": 93 },
      { "label": "Core 12", "temp_input": 20, "temp_max": 93 }
    ]
  }
]
```

Notice that there are now arrays within arrays, and the innermost keys no longer have arbitrary numbers inside them. To obtain this type of JSON today, a consumer could transform today's `sensors -j` output via `jq`:
```sh
sensors -j | jq 'to_entries | map( {"chip_label":.key, "adapter":.value.Adapter, "features": (.value | del(.Adapter) | to_entries | map( {"label":.key} + (.value | with_entries(.key |= sub("^(?<pfx>cpu|curr|energy|fan|humidity|in|intrusion|power|pwm|temp)[0-9]+_";"\(.pfx)_"))) ) ) } )'
```
I contend `sensors` should provide a useful JSON structure in the first place.

Taking the proposal a step further, the JSON structure could also separate by feature type, and include the measurement unit. After all, `sensors` re-scales some values, thus its units aren't the same as those given in the [hwmon documentation](https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface). The example then becomes:

```json
[
  {
    "chip_label": "BAT1-acpi-0",
    "adapter": "ACPI interface",
    "voltage_unit": "V",
    "voltage_features": [
      { "label": "in0", "input": 16.583 }
    ],
    "current_unit": "A",
    "current_features": [
      { "label": "curr1", "input": 0.001 }
    ]
  },
  {
    "chip_label": "bnxt_en-pci-3100",
    "adapter": "PCI adapter",
    "temperature_unit": "°C",
    "temperature_features": [
      { "label": "temp1", "input": 45 }
    ]
  },
  {
    "chip_label": "coretemp-isa-0000",
    "adapter": "ISA adapter",
    "temperature_unit": "°C",
    "temperature_features": [
      { "label": "Package id 0", "input": 24, "max": 93 },
      { "label": "Core 0", "input": 21, "max": 93 },
      { "label": "Core 12", "input": 20, "max": 93 }
    ]
  },
  {
    "chip_label": "bnxt_en-pci-3101",
    "adapter": "PCI adapter",
    "temperature_unit": "°C",
    "temperature_features": [
      { "label": "temp1", "input": 45 }
    ]
  },
  {
    "chip_label": "coretemp-isa-0001",
    "adapter": "ISA adapter",
    "temperature_unit": "°C",
    "temperature_features": [
      { "label": "Package id 1", "input": 24, "max": 93 },
      { "label": "Core 0", "input": 18, "max": 93 },
      { "label": "Core 12", "input": 20, "max": 93 }
    ]
  }
]
```

With this JSON structure, calculating something like “average core temperature” is easy: filter to chips with a `.temperature_features` key, filter to features whose `.label` matches `^Core `, look at `.input`. For example, in `jq`:
```sh
proposed_json | jq -r 'map(.temperature_features[]? | select(.label | test("^Core ")) | .input) | add/length
```

Finally, since JSON output is intended for programmatic consumption, `sensors` could just spit out compact JSON (no unnecessary whitespace/newlines). Formatted output is needed only when humans are designing/studying/debugging intermediate stages of a pipeline, in which case it is easy enough to write `sensors -j | jq` (which furthermore includes colored syntax-highlighting when output goes to a terminal).

In summary, here are the topics for discussion:

1. Should the JSON output of `sensors` be improved?
2. Which is better, the “at a minimum” version given near the start, or the subsequent “taking things a step further” version (the one that includes measurement units)? Or something else?
3. Should JSON output be compact (no unnecessary whitespace/newlines)? (Humans can pipe through `jq` for pretty output.)
4. For backwards compatibility with toolchains that might rely on today's JSON output, should plain `-j` continue producing the original output, while some new option (say `-jj` or `-J` or even `--format=json`) produces proposed output?

**Comment by hramrach** [comment]
There is now a -J option with hopefully more structured data.
